frontend/src/views/instructor/ddahs/import.tsx

```diff
@@ -0,0 +1,319 @@
+
+    // Recompute the diff every time the file changes
+    React.useEffect(() => {
+        // If we have no file or we are currently in the middle of processing another file,
+        // do nothing.
+        if (!fileContent || inProgress) {
+            return;
+        }
+        try {
+            setProcessingError(null);
+            // normalize the data coming from the file
+            const data = normalizeDdahImports(fileContent, applicants);
+
+            // Compute which DDAHs have been added/modified
+            const newDiff = diffImport.ddahs(data, { ddahs, assignments });
+            const newDdahs = newDiff
+                .filter((diff) => diff.status === "new")
+                .map((diff) => diff.obj);
+            const modifiedDdahs = newDiff
+                .filter((diff) => diff.status === "modified")
+                .map((diff) => diff.obj as Ddah);
+            const ddahUpdates = modifiedDdahs.map((modifiedDdah) => ({
+                updatedDdah: modifiedDdah,
+                oldDdah: ddahs.find(
+                    (ddah) => ddah.id === modifiedDdah.id
+                ) as Ddah,
+            }));
+            setNewDdahs(newDdahs);
+            setDdahUpdates(ddahUpdates);
+            setDiffed(newDiff);
+        } catch (e) {
+            console.warn(e);
+            setProcessingError(e);
+        }
+    }, [fileContent, ddahs, assignments, applicants, inProgress]);
+
+    async function onConfirm() {
+        if (!diffed) {
+            throw new Error("Unable to compute an appropriate diff");
+        }
+        const changedDdahs = getChanged(diffed);
+
+        await dispatch(upsertDdahs(changedDdahs));
+
+        setFileContent(null);
+    }
+
+    return (
+        <ImportActionButton
+            onConfirm={onConfirm}
+            onFileChange={setFileContent}
+            dialogContent={
+                <DialogContent
+                    processingError={processingError}
+                    newDdahs={newDdahs}
+                    ddahUpdates={ddahUpdates}
+                    fileLoaded={!!diffed && !inProgress}
+                />
+            }
+            setInProgress={setInProgress}
+            disabled={disabled}
+        ></ImportActionButton>
+    );
+}
+
+const DialogContent = React.memo(function DialogContent({
+    processingError,
+    newDdahs,
+    ddahUpdates,
+    fileLoaded,
+}: {
+    processingError: string | null;
+    newDdahs: (Ddah | Omit<Ddah, "id">)[];
+    ddahUpdates: DdahUpdate[];
+    fileLoaded: boolean;
+}) {
+    if (processingError) {
+        return <Alert variant="danger">{"" + processingError}</Alert>;
+    }
+
+    if (!fileLoaded) {
+        return <p>No data loaded.</p>;
+    }
+
+    if (newDdahs.length === 0 && ddahUpdates.length === 0) {
+        return (
+            <Alert variant="warning">
+                No difference between imported and existing DDAHs.
+            </Alert>
+        );
+    }
+
+    return (
+        <>
```

Comment:
Write out `React.Fragment`. It makes it easier for the autocomplete.

Reply:
Not too sure I understand why, but ok, I'll change it. I don't have a preference either way.